speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/FileDataStorage.cpp

```diff
@@ -98,64 +121,307 @@ void FileDataStorage::Shutdown()
+/// @param p_decisionType the current decision type
+/// @param p_decisionMade if there has already been a decision made
+template <typename TNumber>
+void FileDataStorage::WriteDecision(TNumber p_decision, const std::string& p_decisionType, bool& p_decisionMade)
+{
+    if (p_decision != -1)
     {
-        WRITE_STRING(m_outputStream, "GearDecision");
-        WRITE_VAR(m_outputStream, p_decisions.GetGear());
+        if (!p_decisionMade)
+        {
+            p_decisionMade = true;
+            WRITE_STRING(m_outputStream, "Decisions");
+        }
+        WRITE_STRING(m_outputStream, p_decisionType);
+        WRITE_VAR(m_outputStream, p_decision);
     }
-    if (p_decisions.ContainsLights())
+}
+
+/// @brief Add the new value to the total of the current time steps
+/// @param p_total The current total of the variable
+/// @param p_value The new value of this timestep for the variable
+void FileDataStorage::AddForAveraging(float& p_total, float p_value)
+{
+    p_total += p_value;
+}
+
+/// @brief Calculates the average of the past time steps for a variable and resets it
+/// @param p_total The current total of the variable
+/// @return The average of the past time steps for a variable
+float FileDataStorage::GetAverage(float& p_total) const
+{
+    float average = p_total / static_cast<float>(m_compressionRate);
+    p_total = 0;
+    return average;
+}
+
+/// @brief Add the new value to the array in the correct compression step
+/// @param p_values Array with values from the current compression step
+/// @param p_value The new value of this time step for the variable
+/// @param p_compressionStep The current compression step
+template <typename TNumber>
+void FileDataStorage::AddToArray(TNumber p_values[], TNumber p_value, unsigned long p_compressionStep) const
+{
+    int placeInArray = static_cast<int>(p_compressionStep % static_cast<unsigned long>(m_compressionRate));
+    p_values[placeInArray] = p_value;
+}
+
+/// @brief Get the median of the current compression step using Randomized QuickSelect
+/// @param p_values Array with values from the current compression step
+/// @return The median of the past time steps for a variable
+float FileDataStorage::GetMedian(float* p_values)
+{
+    float startPartition = -1, median = -1;
+    GetMedianUtil(p_values, 0, m_compressionRate - 1, static_cast<int>(m_compressionRate / 2), startPartition, median);
+    return median;
+}
+
+/// @brief Utility function to recursively get the median
+/// @param p_values Array with values from the current compression step
+/// @param p_start start of the current section of the array
+/// @param p_end end of the current section of the array
+/// @param p_middle middle of the current section of the array
+/// @param p_a number before the median of p_values
+/// @param p_b median of p_values
+void FileDataStorage::GetMedianUtil(float* p_values, int p_start, int p_end, int p_middle, float& p_startPartition, float& p_endPartition)
+{
+    if (p_start <= p_end)
```

Comment:
Turn around for less indentation

Reply:
also nothing happens when p_start > p_end and that should be a mistake right? if it should never happen you can use assert(p_start <= p_end) to ensure in debug mode that this is always the case. After that you can remove the identation aswell